wikimedia/wikipedia-ios
# Merge sequential small event only sections into ranged section

**Phabricator:** https://phabricator.wikimedia.org/T265670

### Notes
Merges and collapses sequential sections with small changes matching this criteria:
- The section contains only small events
- Those small events are already compressed into one small event (so the `Section`'s `TypedEvent` count is 1)
- The sections being merged are sequential neighbors – even though their dates might not be strictly sequential, there are no intervening sections between them. For example: M1, M2, M3 where M is a matching section will be collapsed and merged into a new section M0, but M1, L1, M2 will not be merged into a new section.

This merging and collapsing behavior happens at two points - first when a `ArticleAsLivingDocViewModel` is constructed from `SignificantEvents` then again when a new page is fetched and those new events are appended to the existing visible sections. When merging sequential small event sections, a date range is also constructed that represents the collapsed range. This will be used in a future PR to help craft the timestamps in the modal.

I find my work here to be terribly inefficient and gross, _but_ it seems to work on a variety of articles in my testing so far without any noticeable performance implications (because the page size for new data is so small) so maybe the inefficiency isn't that bad here... 🤷🏽‍♂️

### Test Steps
1. Observe AAaLD modal without PR
	- Sequential small events within a page of data are not collapsed and merged.
	- Sequential small events at paging join points are not collapsed and merged.
2. Observe AAaLD modal with PR
	- Sequential small events within a page of data are collapsed and merged.
	- Sequential small events at paging join points are collapsed and merged.

Test articles I used often while implementing this were these two, as they had a lot of recent small events:

https://en.wikipedia.org/wiki/A_History_of_the_Birds_of_Europe
https://en.wikipedia.org/wiki/Ornithology

### Screenshots

<img width="694" alt="small_changes" src="https://user-images.githubusercontent.com/5652327/97618232-04440c00-19dc-11eb-87a5-6317a1c6e9e5.png">
